File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& a, int k) {


        int count=0;
        int i, cur_sum=0;
        int n=a.size();
        map<int, int> map;

        for(i=0;i<n;i++)
        {
            cur_sum=cur_sum+a[i];
            if(cur_sum==k)
                count++;
            
            int rem=cur_sum-k;
            if(map.find(rem)!=map.end())
            {
                count=count+map[rem];
            }
            map[cur_sum]+=1;

        }

        return count;

        
    }
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] a, int k) {
        int n=a.length;
        int target=k;
        int sum=0;
        int count=0;
        Map<Integer,Integer> map= new HashMap<>();
        for(int i=0;i<n;i++)
        {
            sum=sum+a[i];
            if(sum==target)
                count++;
            int rem=sum-target;
            if(map.containsKey(rem))
            {
                count=count+map.get(rem);
            }
            if(map.containsKey(sum))
                map.put(sum,map.get(sum)+1);
            else
                map.put(sum,1);
            
        }

        return count;
        
    }
}